### PR TITLE
[FIX] project: remove duplicated "task created"

### PR DESCRIPTION
**Pull request by phenix-factory**
task-2280184

### CODE REVIEW

**Comment by xdu-odoo** [comment]
> In 13.3, I don't see it displayed on message template. Don't remember if there was a motivation to display it in 13.4 or not, @seb-odoo or @xdu-odoo?

Don't remember well but I think it was due to some empty messages that was displayed

**Comment by alexkuhn** [comment]
> > In 13.3, I don't see it displayed on message template. Don't remember if there was a motivation to display it in 13.4 or not, @seb-odoo or @xdu-odoo?
> 
> Don't remember well but I think it was due to some empty messages that was displayed

Possibly these messages were intended to stay hidden, like how we recently removed user notification messages.
I think the subtype description is intended for being displayed in a config menu to manage subtypes, I don't think it's intentional to have it displayed on message.

**Comment by seb-odoo** [comment]
I already investigated this one and gave this conclusion somewhere else but I can't remember where:

The helper of the `description` field in `mail.message.subtype` literally states that :

> Description that will be added in the message posted for this subtype. If void, the name will be added instead.

This is working as intended.


The problem is the hardcoded and not translatable `body` containing "[record] created" that is added automatically at the creation of records.

The "created" message should have the proper "created subtype" with the proper "created description", that will depend on the user language seeing it. Kinda similar to tracking. And not have an hardcoded `body`.

**Comment by seb-odoo** [comment]
It was there: https://github.com/odoo/odoo/pull/54291

**Comment by alexkuhn** [comment]
It looks off to have subtype description below `mt.comment` messages, e.g. on next.odoo.com (13.4):
message-subtype

I still think subtype description is technical information, to be used in very specific cases like in configuration menu.
It should not be displayed in the message content.

The problem with translation should be solved precisely to the messages that have this problem. Something like having the "body" in formatter returning the content in the language of the current partner.

**Comment by seb-odoo** [comment]
`description` should be removed from the types for which we don't want to display it. I have no idea why it is set on the prod on `mt.comment` (probably due to a lack of migration), but data in 13.4 are not setting this field on `mt.comment` or other basic subtypes.

**Comment by seb-odoo** [comment]
The only "hacky" solution that I would accept would be to hide `description` if `body` is set, keeping in mind this is a problem of either a bad `body` or a bad `description` set at the same time in the first place.

**Comment by alexkuhn** [comment]
> `description` should be removed from the types for which we don't want to display it. I have no idea why it is set on the prod on `mt.comment` (probably due to a lack of migration), but data in 13.4 are not setting this field on `mt.comment` or other basic subtypes.

When I post a new message on next.odoo.com, it shows the subtype description. So doesn't seem just lack of migration on existing messages.

**Comment by seb-odoo** [comment]
No I mean prod had `description` set on `mt.comment` million years ago, and when it was removed from mail data, no migration was made to remove it from our data, because nobody noticed it did anything (since it was broken and didn't, until now).

**Comment by alexkuhn** [comment]
> The only "hacky" solution that I would accept would be to hide `description` if `body` is set, keeping in mind this is a problem of either a bad `body` or a bad `description` set at the same time in the first place.

I don't get why you want the subtype description in the content of the message... The whole purpose of subtype is make partitions on all messages (the "mt.comment" ones, the "mt.log" ones, etc...). Subtype description is some text to explain what this partition means, which is something technical. It's stupid to have it in footer of each message...

```
----
Message 1
<This is a description>
----
Message 2
<This is a log>
----
Message 3
<This is a message>
----
```

You get the idea... We don't want to have this displayed as text every single time. `mt.comment` are just basic messages, so no UI particularities. `mt.log` should just have background (and maybe "Noted by" in the header).

Reliance on subtype description looks a bad hack, we should do better.

**Comment by seb-odoo** [comment]
As I have said before, I think it should appear because it's literally what the field helper states that it does...

> Description that will be added in the message posted for this subtype.

It's clear this feature is used similarly to tracking, to have a custom message that is not a hardcoded `body` in those cases.
It's very clear when looking at examples of `mail.message.subtype` that do have a `description` set. The goal is to show that text as a tracking message.

Earliest occurrence that I could find is this commit: https://github.com/odoo/odoo/commit/bfea9dad64ed500ac471a196a971ea0626a3a97a

**Comment by alexkuhn** [comment]
> As I have said before, I think it should appear because it's literally what the field helper states that it does...
> 
> > Description that will be added in the message posted for this subtype.
> 
> It's clear this feature is used similarly to tracking, to have a custom message that is not a hardcoded `body` in those cases.
> It's very clear when looking at examples of `mail.message.subtype` that do have a `description` set. The goal is to show that text as a tracking message.
> 
> Earliest occurrence that I could find is this commit: [bfea9da](https://github.com/odoo/odoo/commit/bfea9dad64ed500ac471a196a971ea0626a3a97a)

Hm... ok, so the problem is in the misuse of this subtype description for ages... That's annoying... 😒

**Comment by phenix-factory** [comment]
Okey !

Archeology is always interesting, but what can we do exactly ?
I "fixed" a similar bug 2 month ago (https://github.com/odoo/odoo/pull/53252) doing the same thing in this PR.

The subtype table seams to contain only technical information except for the description field. Do we really need this field ?
One can argue that it could avoid a metric ton of "task created" body inside the mail_message table, but in that case why do we fill the body in the first place ?
This also look like an useless optimization since you still need an entry inside the mail_message table, filling the body isn't so bad.

**Comment by seb-odoo** [comment]
Filling body is bad because it is not translatable. The message will forever be with the language of the person that created the record (which can even be the system), and not the language of the user viewing it, which can lead to inconsistencies/message not understandable at all for a user.

Filling body is also bad because it is an html field which requires sanitize, which is slow, and that significantly impacts the time of creation of records, especially perceptible when creating many records in batch.

Finally using body for this kind of message is semantically bad, because it becomes harder to differentiate meta messages such as this one, which is just used for tracking, from actual messages. This is partially resolved by having a proper subtype, but not all "... created" messages have a proper subtype.

**Comment by phenix-factory** [comment]
>Filling body is bad because it is not translatable. The message will forever be with the language of the person that created the record (which can even be the system), and not the language of the user viewing it, which can lead to inconsistencies/message not understandable at all for a user.

I don't understand why the description field is more translatable than the body ?

>Finally using body for this kind of message is semantically bad, because it becomes harder to differentiate meta messages such as this one, which is just used for tracking, from actual messages. This is partially resolved by having a proper subtype, but not all "... created" messages have a proper subtype.

Well since it's the subtype purpose shouldn't we fixe all of them ? I get it could be a lot of work.

I have a few other remarks, looking at the `mail_message_subtype` table:

* `description` isn't a good name, it look optional and "technical". Some suggestions: `forced_body`, `body`, `body_substitute`.
* `description` should replace the body field (by frontend means ?).

What do you think ?

**Comment by seb-odoo** [comment]
Technically: `description` has `translate=True`, `body` does not. Functionally: `description` is a system data that is going to be translated in the first place, `body` is a message that someone posted once in a specific language.

Updating/fixing all of them: good for a master task, can't do as a bugfix in 13.4 due to requiring change in db, migration, etc.

I think the name is fine though, ideally if a message has both a `description` through the subtype and a `body` both should be displayed.

As I already said the ideal solution is to not set the "body" in the first place when it is a "duplicate" of the subtype.
Due to the complexity of it, I would consider out of scope of current fix to fix the python and especially to fix the existing data with a migration.

The only thing I would do for this task is to simply hide duplicate info in JS.
1. The safest option is to hide the description if the body has the same textual value. This might not cover all cases though, especially due to translation stuff, or minor string variations between the two.
2. A bit less safe option but that should be fine is to hide the description if there is any non-empty body set (unfortunately losing the translation of the description in the process). Losing the description in that case is fine because it's fine to assume if a custom body was set it is more important.
3. A bit more risky option but that would be ideal for translation is to do the opposite: to hide the body if there is a description set. This is more unsafe because it assumes nobody has set an "interesting" body along a subtype with a description.

File: addons/mail/static/src/components/message/message_tests.js [hunk 1]
```diff
@@ -738,6 +738,48 @@ QUnit.test('prevent attachment delete on non-authored message', async function (
...
+        id: 100,
+        subtype_description: 'hello',
+    });
+    await this.createMessageComponent(message);
+    assert.containsOnce(
+        document.body,
+        '.o_Message_content',
+        "message should have content"
+    );
+    assert.strictEqual(document.querySelector(`.o_Message_content`).textContent,
```
Comment (alexkuhn) [inline addons/mail/static/src/components/message/message_tests.js:777]: 1st parameter should be on its own line

**Review by alexkuhn — COMMENTED** [review]
Looks good. If you agree with single comment, I can quickly make the changes.

**Review by alexkuhn — APPROVED** [review]
@robodoo r+

**Pull request closed without merging**